lib/web_ui/lib/src/engine/view_embedder/flutter_view_manager.dart

```diff
@@ -101,5 +110,49 @@ class FlutterViewManager {
     // close the streams.
     _onViewCreatedController.close();
     _onViewDisposedController.close();
+
+    // Remove the focus tracking event listeners.
+    domDocument.body?.removeEventListener(_focusin, _focusChangeHandler, true);
+    domDocument.body?.removeEventListener(_focusout, _focusChangeHandler, true);
   }
+
+  int? _latestFocusedViewId;
+
+  // TODO(tugorez): Add handlers for backwards focus.
+  late final DomEventListener _focusChangeHandler = createDomEventListener((DomEvent event) {
+    final DomElement? activeElement = domDocument.activeElement;
+    if (activeElement == null) {
+      return;
+    }
+
+    final int? viewId = _findParentViewIdOrNull(activeElement);
+    if (viewId == _latestFocusedViewId) {
+      return;
+    }
+
+    if(viewId == null) {
+      _dispatcher.invokeOnViewFocusChange(
+        ViewFocusEvent(viewId: _latestFocusedViewId!, viewFocusState: ViewFocusState.unfocused),
+      );
+      _latestFocusedViewId = null;
+    } else {
+      _dispatcher.invokeOnViewFocusChange(
+        ViewFocusEvent(viewId: viewId, viewFocusState: ViewFocusState.forwardFocused),
+      );
+      _latestFocusedViewId = viewId;
```

Comment:
Ohhhh that didn't cross my mind. Great catch! 

Done!

Reply:
I noticed that `PlatformDispatcher.requestViewFocusChange()` is still a [no-op](https://github.com/flutter/engine/blob/266d5d0b5588c1458b7870a14a1f2ccc46a5d931/lib/web_ui/lib/src/engine/platform_dispatcher.dart#L247). Does this mean updating `_lastFocusedViewId` will happen in a future PR?

Reply:
> Does this mean updating _lastFocusedViewId will happen in a future PR?

Sorry I'm not sure I follow the intent here. FWIW it is already being updated. `_lastFocusedViewId` helps this function compute "what changed" with respect to the last event. 

| Last focused view id | Currently focused view id | Action
| --------------------------- | --------------------------------- | -----------------------------  |
| null                            | null                                   | no op                           |
| null                            | 1                                       | view 1 was focused     |  
| 1                                | null                                   | view 1 was unfocused |
| 1                                | 2                                      | view 1 was unfocused, view 2 was focused |

Reply:
When Flutter calls `PlatformDispatcher.requestViewFocusChange()`, does that lead to a `focusin`/`focusout` event in the browser? If yes, then ignore what I said.

If no, then `_lastFocusedViewId` will be out of sync with the real view that's currently focused.

Reply:
Ohhh I see what you meant. 

> does that lead to a focusin/focusout event in the browser? If yes, then ignore what I said.

Yes, it will directly manipulate the dom, wich will prolly trigger focus events. It won't use _lastFocusedViewId at all (I think).

Reply:
Let's take the following sequence:

- `view1` receives focuse, `_lastFocusedViewId` is set to `1`.
- Flutter requests focus for `view2` by calling `PlatformDispatcher.requestViewFocusChange(2)`.
    - If nobody updates `_lastFocusedViewId`, then at this point, it's out of sync (the real focused view is `2`, but `_lastFocusedViewId` is still `1`).
 - Now let's say `view1` receives focus again, `_focusChangeHandler` is called, but it does nothing, because it thinks `view1` is already focused (due to the out of sync `_lastFocusedViewId`).

<hr>

Obviously, if `PlatformDispatcher.requestViewFocusChange()` leads to a `focusin/focusout` event, then `_lastFocusedViewId` will be automatically updated to the correct value.


So, I guess the gist of my comments is: we should make sure that `PlatformDispatcher.requestViewFocusChange()` does actually produce a `focusin/focusout` event so that everything updates correctly.

Reply:
Understood. Agree with everything you said.